File: backend/hermes_client.py
import os
from collections.abc import Mapping

import aiohttp
from fastapi.responses import JSONResponse, Response

HERMES_API_URL = os.environ.get("HERMES_API_URL", "http://127.0.0.1:8642").rstrip("/")


def hermes_headers(*, accept: str = "application/json") -> dict[str, str]:
    headers = {"Accept": accept}
    api_key = os.environ.get("API_SERVER_KEY") or os.environ.get("HERMES_API_KEY")
    if api_key:
        headers["Authorization"] = f"Bearer {api_key}"
    return headers


def upstream_unavailable(exc: Exception) -> JSONResponse:
    return JSONResponse(
        status_code=503,
        content={
            "detail": "Hermes Agent is unavailable",
            "code": "hermes_unavailable",
            "error": str(exc),
        },
    )


async def proxy_json_request(
    method: str,
    path: str,
    *,
    params: Mapping[str, str] | list[tuple[str, str]] | None = None,
    payload: object | None = None,
) -> Response:
    timeout = aiohttp.ClientTimeout(total=30)
    try:
        async with (
            aiohttp.ClientSession(timeout=timeout) as session,
            session.request(
                method,
                f"{HERMES_API_URL}{path}",
                params=params,
                json=payload,
                headers=hermes_headers(),
            ) as upstream,
        ):
            content = await upstream.read()
            response_headers: dict[str, str] = {}
            content_type = upstream.headers.get("Content-Type")
            if content_type:
                response_headers["Content-Type"] = content_type
            session_id = upstream.headers.get("X-Hermes-Session-Id")
            if session_id:
                response_headers["X-Hermes-Session-Id"] = session_id
            return Response(
                content=content,
                status_code=upstream.status,
                headers=response_headers,
            )
    except (aiohttp.ClientError, TimeoutError) as exc:
        return upstream_unavailable(exc)

File: server/hermes-client.ts
import type { ServerConfig } from "./config.js";

export const hermesHeaders = (
  config: ServerConfig,
  accept = "application/json",
): Headers => {
  const headers = new Headers({ Accept: accept });
  if (config.hermesApiKey) {
    headers.set("Authorization", `Bearer ${config.hermesApiKey}`);
  }
  return headers;
};

export const hermesUrl = (
  config: ServerConfig,
  pathname: string,
  query?: string,
): string => `${config.hermesApiUrl}${pathname}${query ? `?${query}` : ""}`;

export const unavailableResponse = (error: unknown): Response =>
  Response.json(
    {
      detail: "Hermes Agent is unavailable",
      code: "hermes_unavailable",
      error: error instanceof Error ? error.message : String(error),
    },
    { status: 503 },
  );

export const proxyHermes = async (
  config: ServerConfig,
  method: string,
  pathname: string,
  options: { query?: string; payload?: unknown; timeoutMs?: number } = {},
): Promise<Response> => {
  try {
    const upstream = await fetch(hermesUrl(config, pathname, options.query), {
      method,
      headers: {
        ...Object.fromEntries(hermesHeaders(config)),
        ...(options.payload === undefined
          ? {}
          : { "Content-Type": "application/json" }),
      },
      body:
        options.payload === undefined
          ? undefined
          : JSON.stringify(options.payload),
      signal: AbortSignal.timeout(options.timeoutMs ?? 30_000),
    });
    const headers = new Headers();
    const contentType = upstream.headers.get("Content-Type");
    const sessionId = upstream.headers.get("X-Hermes-Session-Id");
    if (contentType) headers.set("Content-Type", contentType);
    if (sessionId) headers.set("X-Hermes-Session-Id", sessionId);
    return new Response(upstream.body, { status: upstream.status, headers });
  } catch (error) {
    return unavailableResponse(error);
  }
};

export const safeJsonBody = async (request: Request): Promise<unknown> => {
  try {
    return await request.json();
  } catch {
    return {};
  }
};
